### DISCUSSION TITLE
Question about mesh networking support

### DISCUSSION THREAD

**Discussion by GregoryVPerry** [markqvist/Reticulum#185] [General]
So in terms of building a multi-region mesh with LoRa using Reticulum, is there any current infrastructure in place in terms of connecting regional meshes that are too far apart? This is an option with MQ on Meshtastic but not a very flexible design.

I am currently looking at using the HAL from SX1303 for this, to act as an exit node of sorts for messages that cannot be delivered regionally based on distance.

**Comment by markqvist** [comment]
Yes, absolutely, stuff like this is really where Reticulum shines. You can interconnect any number of physically separate networks over any kind of medium. Reticulum is both a protocol for local, independent networks, and an inter-net protocol.

The different regions don''t even need to know about each other in advance, and you don't have to pre-coordinate anything for it to work.

In practice, the easiest thing is probably to have one or two Reticulum Transport Instances in each region, that are connected over the Internet to the Transport Instances of the other regions. You could set up a single "hub" accessible to all regions over the Internet, or connect them all to each other directly.

Or if you want to avoid the Internet all together, supply your own physical connections between the regions with whatever you can, and just add those connections to one or more of the transport instances in each region.

Once there is some sort of connectivity, physical or virtual, Reticulum handles the rest, and all endpoints have full connectivity to all other endpoints.

**Reply by ricochet-mobi** [reply]
So the SX1303 is an interesting part, what we are working on is a global network based on LoRa, and with an XMR rollout but with POW being exit node services. Reason why SX1303 is because Helium just deprecated their PHAT v1 line which is SX1303 for IOT network; the thought would be to retrofit these SX1303s as exit nodes because that would just require an OS switch for thousands if not tens of thousands of v1 Helium miners that are already installed an connected to the Internet. From my understanding of the part it can capture all 8 spreading sequences and 10 channels at a time, so so method of using the SX1303 HAL to dump Reticulum network traffic that is intended for geographically disparate regional meshes?

**Reply by ricochet-mobi** [reply]
e.g. this device: https://store.rakwireless.com/products/rak5146-kit

tens of thousands of these are all over the place and getting cut from Helium soon.

**Reply by markqvist** [reply]
Can you explain in a bit more detail what you are actually trying to build? The term *exit node* doesn't really make sense within the context of Reticulum, since it converges everything it can reach, into one unified data transport fabric. So from a "Reticulum perspective", there's no entry and exit, there is just "the Ret".

The concept might make sense within the model of your system though, but to understand it properly and be able to offer any kind of advice, I need to know a bit more.

Is it a simple "one-shot" messaging service? Or something a bit more intelligent like [LXMF](https://github.com/markqvist/lxmf)? Or something entirely else?

**Reply by markqvist** [reply]
Sorry, I forgot you had actually already detailed a lot of this in one of your messages to me before making these posts. I've had my head a bit full lately, and way too much communication to keep track of ;)

Either way, I think you should look into LXMF, and try to see if it would fit into what you're trying to do, which I am pretty sure it will. Your system will then also benefit from interoperability with apps like [Sideband](https://unsigned.io/sideband/) and [Nomad Network](https://unsigned.io/nomad_network/), as well as any future apps and systems based on LXMF.

You also get all of the "exit node" and offline message storage functionality for free, basically, since LXMF handles this very gracefully.

Sending a message to someone is just a few lines of code with LXMF and Reticulum, so you just need to worry about the specifics of your own system :)

**Reply by GregoryVPerry** [reply]
Ok, so the concept is pretty simple. We've got a dual LoRa device we are launching with integrated GPS capabilities, TFT high resolution display, HID emulation for PS/1 keyboard cold wallet / password manager capabilities via BLE or USB-C. $99, all open source including Gerbers, schematics, firmware etc. 700mah rechargeable battery, self contained, a little larger than a business card in terms of dimensions. Pocket-sized mobile communicator, used for 100% globally free, cryptographically secure, anonymous (semi-instant) messages and blockchain transactions, nothing more.

Goal is to launch an XMR token, but with POW substituted for exit node services. Meaning, you purchase or build the device, then pair it with an Internet service via BLE or WiFi, and then you participate in equitable distribution of the network's daily POW rewards. So that regional meshes, say for example a group of devices on a college campus, can pass messages to remote addresses/nodes that can't be delivered to local mesh connected addresses. You don't have to pair the device with an Internet connection and can use it like a walkie talkie or even participate in the global network, as long as you are in range of local mesh with exit node participation to route global messages to. But in order to mine using the mobile device (or perhaps an SX1303 basestation as mentioned below), you have to connect it to an Internet source.

So in this context, we have a tiny device we are releasing open source, and which has dual SX1262 LoRa + GPS + BLE + WiFi, based on ESP32-S3. The primary LoRa radio is intended for instant messages and blockchain transactions, second LoRa radio for mesh control channel operations to establish and maintain a constantly-evolving global routing fabric based on primarily GPS device location and secondary coherent SDR direction finding based on LoRa preamble analysis. None of this is set in stone, both LoRa radios could be used concurrently if that would be supported with Reticulum, and with in-band control channel messages. Two of the devices can be used like high-tech walkie talkies, to send instant messages within say 1m local range afforded by point-to-point 915Mhz LoRa (we have multiple versions of the hardware design based on ISM region), but if using Reticulum the assumption is that all nodes running a comparable version of the firmware will pass messages for all other nodes irrespective of destination, in order to establish P2P anonymous meshes for communication. As an incentive, users of the device can simply tether it via BLE to their smartphone or join it with their WiFi network to provide the concept of exit node services, e.g. Internet-based traffic to connect regional LoRa addresses with remote (global) addresses if the message cannot be delivered locally, in order to participate in daily network XMR token awards for POW.

We are releasing an open source Flutter app for this, that will be published to both Google and iOS app stores. This is a background app which provides BLE beaconing and tethering, to pair the device with for mining and which provides Internet connectivity for connecting regional mesh messages to global messages via IPFS pubsub, or comparable method if Reticulum supports it.

There is also a security IC for serialized ECC ops as well as various other sensors, intended for crypto operations including both cold and hot wallet. Regional meshes would be established via these handheld devices, but in order to connect these regional device-based meshes to remote meshes, the concept of an exit node - incentivized by an XMR token daily distribution for providing Internet connectivity - is used to connect a local mesh with neighboring meshes, in order to facilitate global connectivity, and devoid of central operator or authority. Exploring funding a gRPC edge for IPFS pubsub, for message passing beyond locally connected nodes; I have studied a portion of LXMF so far but don't know enough yet of the architecture to determine how Reticulum messaging could be implemented in a completely decentralized fashion, without the requirement for central servers to bootstrap and run. IPFS pubsub could be used for this, but has a nascent sub platform based on REST which requires plumbing and admin access to retrieve subscribed message results.

Also including incentives for attaching KrakenSDR units, which provide for direction finding capabilities using fixed antenna arrays in conjunction with coherent SDR. So that geographical routing capabilities based on both on-device GPS and exit node operator w/ KrakenSDR would be accessible to the network in a similar fashion as KrakenSDR's pro line of global direction finding capabilities.

No founders or ICO for the XMR token launch, just a 36 month equitable distribution divided by all exit nodes, and with increasing difficulty over the fixed 36 month token distribution period in order to build the global mesh, hopefully without exit node requirements at the 36 month mark based on density of the network (obviously excepting continental / geographic divides but in the advent of Starlink Internet connectivity this is less of a concern). At the 36 month mark a network change from POW to POS, and with staked validators distributing a percentage of awards to remaining exit nodes on the network under POS distribution.

Purchasers of the handheld device (called PiNG), will already have the capability of providing Internet (exit node) services via BLE or WiFi. However, Helium has been very profitable, with a similar token distribution method, and there are many thousands if not tens of thousands of these v1 SX1303 devices out there, already connected to the Internet and with a Raspberry Pi onboard. These devices can no longer join the Helium network, but are very powerful in terms of capability (able to capture all spreading factors and 10 channels concurrently), and are already installed and waiting for a different OS distribution to leverage in terms of mining. And they are all sitting there, making less money than the cost of power and Internet connectivity based on changes to the Helium network, which is for-profit in terms of LoRa network access for IoT applications. 

So the thought is, how to distribute a Reticulum-based OS for these Raspberry Pi-based SX1303 Helium v1 devices as stand-alone miners/basestations and with 10X reward (because 10X LoRa channels, could we use the SX1303 HAL to simply capture and pass ambient Reticulum LoRa traffic to the Internet?), so that regional LoRa nodes can be observed and routed for global delivery over the Internet while the global mesh is being established etc

**Reply by markqvist** [reply]
Thanks for your detailed explanation.

Could I ask you, before I give you a more in-depth answer to all of it, how much of this has been formed into actual code, or just a somewhat formal protocol specification, and how much of it is purely at the "idea stage"? Do you have any public code repositories for this anywhere?

**Reply by GregoryVPerry** [reply]
The PiNG hardware design is complete and will be released open source this week once the website is finished. In terms of firmware preliminary LoRa drivers are in place and working, based on MicroPython. Then I ran into Reticulum per our initial email correspondence a few weeks ago.
